Fix issue with Nemesis variable hiding.

jwpeterson opened the pull request:
This update includes two bug fixes (libMesh/libmesh#1704 & libMesh/libmesh#1705) as well as the new `DenseMatrix::outer_product` routine added by @joshuahansel. Since we have training next week and the following week, we should probably either do this now, or wait 2 weeks...

Refs #11469.

jwpeterson committed:
Add test verifying Nemesis variable hiding actually works.

The original version of this test was not sufficient to exercise an
indexing bug in libmesh, the new version of the test is more complete
in this regard. The new test has multiple nonlinear and auxiliary
variables of both nodal and elemental types. I have also tested this
on various combinations of hidden/unhidden vars and aux vars and the
variable hiding code seems to work for all of them.

In MOOSE, what currently happens is that only nodal variables are
written to Nemesis files, so that CONSTANT, MONOMIAL auxes don't show
up in the output at all. This is due to the fact that we automatically
hide all variables, except those in getNodalVariableOutput(), in
Nemesis.C.

There remain two versions of this test which are specific to different
compiler/architecture combos, since Parmetis partitioning is still
known to be architecture dependent (due to different random number
generators being used on different systems).  This test now has its
own input file rather than piggybacking off others.

Refs #11469.

jwpeterson committed:
libMesh update.

Brief summary of changes in this update:
* Fix to recent parallel_sync changes.
* Fix indexing bug in Nemesis variable hiding code.
* Convert all enums to C++11 fixed type enumerations.
* Add std::tuple-enabled versions of various BoundaryInfo functions
  (originals are now deprecated).
  - build_node_list
  - build_edge_list
  - build_active_side_list
  - build_shellface_list
  - build_side_list
* Generalizations to ParmetisPartitioner required for #11629.

Refs #000.

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit bdf4ac0280612bc8e43aa30a0805c45ffc210205
+Subproject commit 4661a00faa9a5837c6ef058449ad9a5f49050802
```
**Caution!** This contains a submodule update

roystgnr commented:
I wouldn't wait on the parallel refactoring.  @pbauman and @bboutkov just told me they can get #1702 to hang on 16 processors on the Buffalo cluster; we don't want to merge until I can see if I can replicate that or not.

I'm tempted to bet on "not", based on previous annals of irreproducible results on their system.  In the phrase "Buffalo cluster" you'd think the adjective comes from "noun: relating to the University of Buffalo" but it's actually from "verb: to outwit, confuse, deceive or intimidate".

However, that's by far the smallest run they've managed to make trigger a weird MPI hang, so even if we can't trigger it elsewhere I want to investigate it before merging.  Failures on a single node mean they can rule out interconnect flakiness, and if the problem is in their software stack then it might be another case like libMesh/libmesh#1699 where the code really was wrong even though popular stacks were letting it slide.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/193465/) on d6b9bcd wanted to post the following:

View the site [here](http://mooseframework.inl.gov/docs/PRs/11565/site)

This comment will be updated on new commits.

jwpeterson committed:
Refactor MooseMesh::buildNodeList() to use new build_node_list() routine.

Refs libMesh/libmesh#1712.

jwpeterson committed:
Refactor MooseMesh::buildBndElemList() to use new build_active_side_list() routine.

Refs libMesh/libmesh#1712.

jwpeterson committed:
Refactor MooseMesh::buildPeriodicNodeSets() to use new build_node_list() routine.

Refs libMesh/libmesh#1712.

jwpeterson committed:
Refactor MooseMesh::ghostGhostedBoundaries() to use new build_side_list() routine.

Refs libMesh/libmesh#1712.

jwpeterson committed:
Add tuple-enabled version of MooseMesh::buildSideList().

The original version will be officially deprecated once I can
extricate its usage from the PenetrationThread class.

Refs libMesh/libmesh#1712.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/197339/) on 9cdff3a wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/docs/PRs/11565/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/docs/PRs/11565/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format c44a7f0d4176f1430019221a946187c59f6f32cc`

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit bdf4ac0280612bc8e43aa30a0805c45ffc210205
+Subproject commit 2620b001f7cfac3af4c37a21a465cb29591876b5
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

jwpeterson committed:
Officially deprecate the original MooseMesh::buildSideList()

This apparently should not cause errors during testing, and we can
officially remove the old usage in a separate PR.

permcody commented on framework/src/mesh/MooseMesh.C:
```diff
@@ -600,51 +600,51 @@ MooseMesh::buildNodeList()
-  for (int i = 0; i < n; i++)
+  int n = bc_tuples.size();
+  _bnd_elems.clear();
+  _bnd_elems.reserve(n);
```
Minor: Again, I realize this is only refactoring but would we be better off with just using `assign` instead of `clear`, `reserve`? Actually this whole loop could be refactored to a slick `for_each` or `transform` call. No changes necessary, just commenting.

jwpeterson replied:
I don't think `assign` is the right thing here, since we are doing `push_back` in the subsequent loop. In fact the `reserve(n)` is probably going to be a no-op in most cases after the first since we are just reusing the same storage in the `_bnd_elems` container. It might be possible to get rid of the whole loop, but I think it would be harder to read overall...

permcody commented on framework/src/mesh/MooseMesh.C:
```diff
@@ -1980,9 +1978,19 @@ MooseMesh::buildSideList(std::vector<dof_id_type> & el,
                          std::vector<unsigned short int> & sl,
                          std::vector<boundary_id_type> & il)
 {
+  // TODO: Officially deprecate this routine when it is no longer being used by PenetrationLocator.
```
Wait, we should be able to deprecate this now. The whole point of the new deprecation system is that it doesn't trigger errors in the TestHarness.

jwpeterson replied:
OK, wasn't sure about this, I will uncomment that.

permcody commented on framework/include/mesh/MooseMesh.h:
```diff
@@ -185,10 +185,16 @@ class MooseMesh : public MooseObject, public Restartable
+  /**
+   * As above, but uses the non-deprecated std::tuple interface.
+   */
+  std::vector<std::tuple<dof_id_type, unsigned short int, boundary_id_type>> buildSideList();
```
👍

permcody commented on test/tests/outputs/variables/tests:
```diff
@@ -37,21 +37,20 @@
+    cli_args = 'Outputs/out/file_base=intel_show_hide_out'
     min_parallel = 2
     max_parallel = 2
     exodiff = 'intel_show_hide_out.e.2.0 intel_show_hide_out.e.2.1'
```
Can you please add the new `requirement`, `design`, and `issues` parameters?

permcody commented on framework/src/mesh/MooseMesh.C:
```diff
@@ -600,51 +600,51 @@ MooseMesh::buildNodeList()
-    _bnd_elem_ids[ids[i]].insert(elems[i]);
+    auto elem_id = std::get<0>(t);
+    auto side_id = std::get<1>(t);
+    auto bc_id = std::get<2>(t);
```
I was going to suggest using `std::tie`, but I don't think I like that syntax as much. It looks like there's no way to use `auto` with it. I think this is probably better.

jwpeterson replied:
Agreed, I think `std::tie` will essentially be obsoleted by the C++17 structured bindings.

jwpeterson commented on framework/src/mesh/MooseMesh.C:
```diff
@@ -600,51 +600,51 @@ MooseMesh::buildNodeList()
+    auto node_id = std::get<0>(t);
+    auto bc_id = std::get<1>(t);
+
+    _bnd_nodes.push_back(new BndNode(getMesh().node_ptr(node_id), bc_id));
```
I thought about this, as it would allow us to drop the manual `delete`s from `freeBndNodes()` and `freeBndElems()`. There should be effectively no additional overhead for `unique_ptr` since we don't provide a custom deleter, so it would be worth looking at.

jwpeterson committed:
Add requirement/design/issues fields for new tests.

jwpeterson commented:
@fdkong The libmesh update in this PR now contains your `ParmetisPartitioner` changes.

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit bdf4ac0280612bc8e43aa30a0805c45ffc210205
+Subproject commit 12a11df94850d0e44884e943e02f049c05183b6f
```
**Caution!** This contains a submodule update

fdkong commented:
> @fdkong The libmesh update in this PR now contains your ParmetisPartitioner changes.

Thanks

jwpeterson commented:
We should probably also include https://github.com/libMesh/libmesh/pull/1722 in this update, although it doesn't really affect any of our configurations.

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit bdf4ac0280612bc8e43aa30a0805c45ffc210205
+Subproject commit ceb1e02bee97b3758bfcdc37cb87bc8954a89030
```
**Caution!** This contains a submodule update
